torchrl/data/utils.py

```diff
@@ -35,6 +43,206 @@
+    if not is_tensor_collection(td):
+        return None if td.numel() == 0 else td.clone()
+
+    td = td.clone()
+
+    if isinstance(td, LazyStackedTensorDict):
+        for td_index in range(len(td.tensordicts)):
+            sub_td = td.tensordicts[td_index]
+            sub_td = relazyfy_keys(sub_td)
+            td.tensordicts[td_index] = sub_td
+
+    for key in list(td.keys()):
+        try:
+            value = relazyfy_keys(td.get(key))
+            if value is None:
+                del td[key]
+            else:
+                td.set(
+                    key,
+                    value,
+                )
+        except RuntimeError as err:
+            if re.match(r"Found more than one unique shape in the tensors", str(err)):
+                pass
+            else:
+                raise err
+
+    if isinstance(td, TensorDict) and not len(td.keys()):
+        return None
+    return td
+
+
+def _empty_like(td, batch_size):
+    if is_tensor_collection(td):
+        return td.empty()
+    else:
+        shape = [dim if i < len(batch_size) else 0 for i, dim in enumerate(td.shape)]
+
+        return torch.empty(
+            shape,
+            dtype=td.dtype,
+            device=td.device,
+        )
+
+
+def check_no_lazy_keys(td, recurse: bool = True):
+    """Given a TensorDictBase, returns true if there are no lazy keys."""
+    if isinstance(td, LazyStackedTensorDict):
+        keys = set(td.keys())
+        for t in td.tensordicts:
+            if recurse and not check_no_lazy_keys(t):
+                return False
+            if set(t.keys()) != keys:
+                return False
+    elif isinstance(td, TensorDict) and recurse:
+        for i in td.values():
+            if not check_no_lazy_keys(i):
+                return False
+    elif isinstance(td, torch.Tensor):
+        return True
+    else:
+        return False
+
+    return True
+
+
+def get_all_keys(td, include_lazy: bool):
+    """Given a TensorDictBase, returns all lazy and not lazy keys as a set tuples."""
+    keys = set()
+    if isinstance(td, LazyStackedTensorDict) and include_lazy:
+        for t in td.tensordicts:
+            keys = keys.union(get_all_keys(t, include_lazy=include_lazy))
+    if isinstance(td, TensorDictBase):
+        for key in td.keys():
+            try:
+                keys.add((key,))
+                value = td.get(key)
+                inner_keys = get_all_keys(value, include_lazy=include_lazy)
+                for inner_key in inner_keys:
+                    keys.add((key,) + _unravel_key_to_tuple(inner_key))
+            except RuntimeError:
+                pass
+    return keys
+
+
+def all_eq(
+    td: Union[TensorDictBase, torch.Tensor],
+    other: Union[TensorDictBase, torch.Tensor],
+):
+    """Returns true if the two classes match all entries in the keys and stack dimensions."""
+    if td.__class__ != other.__class__:
+        return False
```

Comment:
why can't we compare a TensorDict and a LazyStackedTensorDict?

Reply:
the idea was that the 2 td we are comparing should have exactly identical structure.
happy to change it if you want

Reply:
i also made it an arg